Python/Product/Analysis/LanguageServer/IServer.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    public interface IServer {
```

Comment:
Do we need to expose everything? I'd keep interface to absolute minimum. Extension definitely should not be able to call Shutdown.... ;-) If we want extension to be able to intercept calls in the protocol then it should have appropriate hooks we will call (not events since extension will definitely forget to detach). Something like

```cs
interface ILanguageServerExtension {
Task OnIntitialize(CancellationToken t); // we'll call it after Init
Task OnShutdown(...) 
Task<CompletionList> OnCompletion(...)
...
}

We can look at specific list of notifications.

Reply:
You are far too distrustful of extensions :) Even with the interface, they can just cast to `Server` and do whatever they want anyway (or you know, run arbitrary code to pwn your machine right there) - I just added the interface because you all said it would make it easier to write tests.

Detaching from events doesn't matter, as neither Server nor any extensions will significantly out-live each other.

The whole point of being integrated into the server here is to have full access to everything. Proxying is a much better option if you want a thin and protected layer - this is for deep integrations that can and do use everything available.